provider/aws: Manage IAM policy documents

jkinred opened the pull request:
This fixes #5201.

The `aws_iam_role` and `aws_iam_policy` resources don't manage policy documents because the resource never updates the policy after initial creation.
#5201 pertains to `aws_iam_policy`. This change aims to apply the same fix to `aws_iam_role`.

I have chosen to perform a more functional comparison by "normalizing" the policy documents by sorting keys and removing whitespace and newline characters. I have used `encoding/json` functions to perform that work.

jkinred commented:
I am looking for advice on the best way to log and return errors in Terraform/Go. As an example at line 254 of resource_aws_iam_policy.go.

Have I gone about this the wrong way, should I be passing in `schema.ResourceData` to the `normalizePolicyDocument` function and calling `d.Set()`? Although this doesn't look like it would work for `StateFunc`.

stack72 commented on builtin/providers/aws/resource_aws_iam_policy.go:
```diff
@@ -32,6 +33,13 @@ func resourceAwsIamPolicy() *schema.Resource {
 			"policy": &schema.Schema{
 				Type:     schema.TypeString,
 				Required: true,
+				StateFunc: func(v interface{}) string {
```
Can you describe why you are using a StateFunc here?

jkinred replied:
Perfect, I wasn't aware of the existing `normalizeJson` function.

stack72 replied:
luckily this is already used in a tonne of other places :)

stack72 commented on builtin/providers/aws/resource_aws_iam_role.go:
```diff
@@ -128,6 +139,13 @@ func resourceAwsIamRoleReadResult(d *schema.ResourceData, role *iam.Role) error
 	if err := d.Set("unique_id", role.RoleId); err != nil {
 		return err
 	}
+	policy, err := url.QueryUnescape(*role.AssumeRolePolicyDocument)
```
I don't think we need to use a combination of `url.QueryUnescape` + `normalizePolicyDocument` - we should just be able to use the `normalizeJson` func already in terraform.

jkinred replied:
I tried to do this using just the built-in `normalizeJson` function but receive the following error:

```
assume_role_policy: "Error parsing JSON: invalid character '%' looking for beginning of value" => "{\"Statement\":[{\"Action\":\"sts:AssumeRole\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":[\"arn:aws:iam::000000000000:role/ec2-rundeck\",\"arn:aws:iam::000000000000:role/ec2_rundeck\"],\"Service\":\"ec2.amazonaws.com\"},\"Sid\":\"\"}],\"Version\":\"2012-10-17\"}"
```

The policy returned in the underlying response is full of HTML character codes, like so:

```
%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22ec2.amazonaws.com%22%2C%22AWS%22%3A%5B%22arn%3Aaws%3Aiam%3A%3A00000000000%3Arole%2Fec2-rundeck%22%2C%22arn%3Aaws%3Aiam%3A%3A00000000000%3Arole%2Fec2_rundeck%22%5D%7D%2C%22Action%2
2%3A%22sts%3AAssumeRole%22%7D%5D%7D
```

In my latest commit it works nicely with the `url.QueryUnescape` in place. Should I be doing something else with the string before passing it to the `normalizeJson` function? Should the `url.QueryUnescape` be added in the `normalizeJson` function?

I looked at another resource (`resource_aws_cloudformation_stack.go`) where normalizeJson is used and compared the sample responses provided in the AWS API doco, and it doesn't seem to me as though the response should be so much different, so I will have to dig further to work out what the difference is.

The sample responses that I'm looking at are:
http://docs.aws.amazon.com/IAM/latest/APIReference/API_GetRole.html
http://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_GetTemplate.html

jkinred commented:
PR updated to use pre-existing normalizeJson function.

stack72 commented:
hi @jkinred,

this is looking good, the build fails due to this:

```
# github.com/hashicorp/terraform/builtin/providers/aws
builtin/providers/aws/resource_aws_iam_role.go:5: imported and not used: "log"
```

Please fix that up and I will run the tests and make sure it works as expected :)

jkinred commented:
Whoops, fixed that up.

I'm pretty keen to write tests for this too. Can you recommend an existing test which is a good example to base my tests on?

stack72 commented:
@jkinred - nice to hear you want to write tests :)

here is a good example; https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_iam_group_policy_test.go#L14

When you are done, can you rebase this into a few logical commits so that we can keep the commit log clean and descriptive to the changes made: examples are like this:

<img width="624" alt="screen shot 2016-02-25 at 12 58 29" src="https://cloud.githubusercontent.com/assets/227823/13320257/81d63e3c-dbbf-11e5-9a12-ea7baf8794e1.png">

jkinred commented:
Hi @stack72,

I just got back to this one.

Having a think about the tests, I'm not sure how to go about writing a test which covers the use case which this PR is fixing, because we depend on the policy changing externally.

1. Terraform creates `aws_iam_role` resource
1. Role policy is changed outside of Terraform (e.g. AWS Console)
1. Terraform is run and detects difference, re-applying original policy

It doesn't seem likely, but is Step 2 is possible in the acceptance tests?

The existing test should exercise the new code, in that Terraform will retrieve the policy from AWS API using the new code and then perform the comparison.

Previously, Terraform only compared the configuration with the state. Now, Terraform will refresh the state and then do the comparison, which is exactly what we want and I believe the existing test case will cover that.

If you're happy with this I will rebase the commits and finish this PR.

Thoughts?

stack72 commented:
@jkinred you are correct - step 2 isn't possible in this framework. Continue with the rebase and I can run some tests before merging :)

jkinred commented:
Hi @stack72,

I have rebased my changes and also added a test for `aws_iam_policy` (my comments about tests above were in relation to `aws_iam_role`).

Unfortunately, I am seeing a problem with the acceptance test for `aws_iam_role` when running locally:
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSRole_basic'
--- FAIL: TestAccAWSRole_basic (22.91s)
        testing.go:148: Step 0 error: After applying this step and refreshing, the plan was not empty:

                DIFF:

                UPDATE: aws_iam_role.role
                  assume_role_policy: "{\"Statement\":[{\"Action\":\"sts:AssumeRole\",\"Effect\":\"Allow\",\"Principal\":{\"Service\":\"ec2.amazonaws.com\"}}],\"Version\":\"2012-10-17\"}" => "{\"Statement\":[{\"
Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"ec2.amazonaws.com\"]}}],\"Version\":\"2012-10-17\"}"

                STATE:

                aws_iam_role.role:
                  ID = test-role
                  arn = arn:aws:iam::586578793606:role/test-role
                  assume_role_policy = {"Statement":[{"Action":"sts:AssumeRole","Effect":"Allow","Principal":{"Service":"ec2.amazonaws.com"}}],"Version":"2012-10-17"}
                  name = test-role
                  path = /
                  unique_id = AROAJUWTPHS3KCNDKBFGG
```

I can see that the difference is with some of the attributes in the `assume_role_policy` attribute document being turned into lists: `\"Service\":[\"ec2.amazonaws.com\"]`.

Interestingly, when I extract the following policy from the test:
```
resource "aws_iam_role" "role_update_test" {
  name = "tf_new_name"
  path = "/test/"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF
}
```

...and `terraform apply` and then `terraform plan -refresh=false` there is no change detected:
```
jonathan@nomad $ terraform apply
aws_iam_role.role_update_test: Creating...
  arn:                "" => "<computed>"
  assume_role_policy: "" => "{\"Statement\":[{\"Action\":\"sts:AssumeRole\",\"Effect\":\"Allow\",\"Principal\":{\"Service\":\"ec2.amazonaws.com\"},\"Sid\":\"\"}],\"Version\":\"2012-10-17\"}"
  name:               "" => "tf_new_name"
  path:               "" => "/test/"
  unique_id:          "" => "<computed>"
aws_iam_role.role_update_test: Creation complete

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate
[Linux][~/workspace/golang/src/github.com/hashicorp/terraform/testcase]
jonathan@nomad $ terraform plan -refresh=false
No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

Seems like some kind of internal state issue? I've put a smattering of `log.Printf` statements about but wasn't able to nail it down. Any ideas?

stack72 commented:
Hi @jkinred - I am going to pull your branch down and have a look at this. Nice work on adding a test though :)

fmasuhr commented:
@stack72 what is the state of this? I would like to see this merged since this is a huge issue regarding the management of policies.

jen20 commented:
Hi @jkinred! I'm going to pick up this pull request now and try to get it merged for Terraform 0.7.2. Thanks for your work, and sorry it's taken so long to review!

jen20 commented:
Hi @jkinred! Looking at the test here - it passes without the actual fix being applied, so I'm not strictly sure we're testing the correct thing here. That said I can definitely see the issue here, and will look at fixing it up shortly.

dtolnay commented:
@jen20: is there anything that @jkinred or I could do to move this along? This fix would be a big help for me.

Ninir commented:
Hey @jkinred 

Do you think you could migrate this to github.com/terraform-providers/terraform-provider-aws (which is now the AWS provider reference repository) please?

This would allow us to fix this quickly :)

Thanks!

apparentlymart commented:
Hello @jkinred, and thanks for working on this!

As part of the the Terraform 0.10 release earlier this year, all of the Terraform providers were moved to their own repositories in [the terraform-providers GitHub organization](https://github.com/terraform-providers), and removed from the Terraform Core repository.

Unfortunately due to the fact that new issues and pull requests are being opened constantly, it was not possible for the various provider maintainers to merge all outstanding pull requests before this split, and there is no automatic way to migrate a pull request to a new repository.

As a result, this pull request can sadly no longer be applied as-is, and so I'm going to close it.

If you or someone else has the time and motivation to apply same changes to [the `aws` provider repository](https://github.com/terraform-providers/terraform-provider-aws) and open a new PR there, the maintainers of that provider should be able to review and merge it.

Thanks again for working on this, and sorry it was not able to be merged before the provider repository changes.
